core/src/main/java/com/linecorp/armeria/internal/common/stream/DecodedHttpStreamMessage.java

```diff
@@ -242,7 +250,11 @@ public void onNext(HttpObject obj) {
                 } else {
                     // Handler didn't produce anything, which means it needs more elements from the upstream
                     // to produce something.
-                    askUpstreamForElement();
+                    if (demand() > 0) {
```

Comment:
This will make test failed due to other codes may want decoder try to decode until it gets the item. But it means the BodyPart' subscribe will get all the data even we just request 1 `HttpData`. Maybe need help about this part. 🙇

Reply:
Try to fix this by 
```
diff --git a/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java b/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java
index 625fd077f..7f47c7ab9 100644
--- a/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java
+++ b/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java
@@ -249,8 +249,6 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
     }

     private void doRequest(long n) {
-        onRequest(n);
-
         final long oldDemand = demand;
         if (oldDemand >= Long.MAX_VALUE - n) {
             demand = Long.MAX_VALUE;
@@ -258,6 +256,8 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
             demand = oldDemand + n;
         }

+        onRequest(n);
+
         if (oldDemand == 0 && !queue.isEmpty()) {
             notifySubscriber0();
         }
```

Because the `onRequest` will cause `onNext` works, and I want it also check the `demand` in implementation. Not sure if it's make sense or not.